google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1alpha2/JsonStreamWriter.java

```diff
@@ -0,0 +1,359 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.bigquery.storage.v1alpha2;
+
+import com.google.api.core.*;
+import com.google.api.gax.batching.BatchingSettings;
+import com.google.api.gax.core.CredentialsProvider;
+import com.google.api.gax.core.ExecutorProvider;
+import com.google.api.gax.retrying.RetrySettings;
+import com.google.api.gax.rpc.*;
+import com.google.cloud.bigquery.storage.v1alpha2.Storage.AppendRowsRequest;
+import com.google.cloud.bigquery.storage.v1alpha2.Storage.AppendRowsResponse;
+import com.google.common.base.Preconditions;
+import com.google.protobuf.Descriptors;
+import com.google.protobuf.Descriptors.Descriptor;
+import java.io.IOException;
+import java.util.*;
+import java.util.logging.Logger;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * A StreamWriter that can write JSON data (JSONObjects) to BigQuery tables. The JsonStreamWriter is
+ * built on top of a StreamWriter, and it simply converts all JSON data to protobuf messages then
+ * calls StreamWriter's append() method to write to BigQuery tables. It maintains all StreamWriter
+ * functions, but also provides an additional feature: schema update support, where if the BigQuery
+ * table schema is updated, the JsonStreamWriter is guaranteed to be updated after some time.
+ */
+public class JsonStreamWriter {
+  private static String streamPatternString =
+      "projects/[^/]+/datasets/[^/]+/tables/[^/]+/streams/[^/]+";
+  private static Pattern streamPattern = Pattern.compile(streamPatternString);
+  private static final Logger LOG = Logger.getLogger(JsonStreamWriter.class.getName());
+
+  private BigQueryWriteClient client;
+  private String streamName;
+  private StreamWriter streamWriter;
+  private Descriptor descriptor;
+  private Table.TableSchema BQTableSchema;
+
+  /**
+   * Constructs the JsonStreamWriter
+   *
+   * @param builder The Builder object for the JsonStreamWriter
+   */
+  private JsonStreamWriter(Builder builder)
+      throws Descriptors.DescriptorValidationException, IllegalArgumentException, IOException,
+          InterruptedException {
+    Matcher matcher = streamPattern.matcher(builder.streamName);
+    if (!matcher.matches()) {
+      throw new IllegalArgumentException("Invalid stream name: " + builder.streamName);
+    }
+
+    this.streamName = builder.streamName;
+    this.client = builder.client;
+    this.BQTableSchema = builder.BQTableSchema;
```

Comment:
This violates Java member variable naming?

Reply:
Changed BQTableSchema to be tableSchema.